File: LeetCode/src/binarysearch/MedianOfTwoSortedArrays.java
package binarysearch;

public class MedianOfTwoSortedArrays {
    public double findMedianSortedArrays(int[] nums1, int[] nums2) {
        if (nums1.length > nums2.length) {
            return findMedianSortedArrays(nums2, nums1);
        }

        int m = nums1.length;
        int n = nums2.length;
        int left = 0, right = m, half = (m + n + 1) / 2;

        while (left <= right) {
            int part1 = (left + right) / 2;
            int part2 = half - part1;

            int part1left = (part1 == 0) ? Integer.MIN_VALUE : nums1[part1 - 1];
            int part1right = (part1 == m) ? Integer.MAX_VALUE : nums1[part1];
            int part2left = (part2 == 0) ? Integer.MIN_VALUE : nums2[part2 - 1];
            int part2right = (part2 == n) ? Integer.MAX_VALUE : nums2[part2];

            if (part1left <= part2right && part2left <= part1right) {
                if ((m + n) % 2 == 0) {
                    return ((Math.max(part1left, part2left) +
                            Math.min(part1right, part2right)) /
                            2.0);
                } else {
                    return Math.max(part1left, part2left);
                }
            } else if (part1left > part2right) {
                right = part1 - 1;
            } else {
                left = part1 + 1;
            }
        }

        return 0.0;
    }

    public static void main(String[] args) {
        MedianOfTwoSortedArrays solution = new MedianOfTwoSortedArrays();
        int[] nums1 = {2, 3, 4};
        int[] nums2 = {1, 2, 5, 7};
        System.out.println(solution.findMedianSortedArrays(nums1, nums2)); // Output: 2.0
    }
}

File: LeetCode_Python/binarysearch/MedianOfTwoSortedArrays.py
class MedianOfTwoSortedArrays:
    def findMedianSortedArrays(self, nums1: list[int], nums2: list[int]) -> float:
        if len(nums1) > len(nums2):
            nums1, nums2 = nums2, nums1

        m, n = len(nums1), len(nums2)
        left, right = 0, m

        while left <= right:
            i = (left + right) // 2
            j = (m + n + 1) // 2 - i

            nums1_left_max = float("-inf") if i == 0 else nums1[i - 1]
            nums1_right_min = float("inf") if i == m else nums1[i]
            nums2_left_max = float("-inf") if j == 0 else nums2[j - 1]
            nums2_right_min = float("inf") if j == n else nums2[j]

            if nums1_left_max <= nums2_right_min and nums2_left_max <= nums1_right_min:
                if (m + n) % 2 == 0:
                    return (
                        max(nums1_left_max, nums2_left_max)
                        + min(nums1_right_min, nums2_right_min)
                    ) / 2
                else:
                    return max(nums1_left_max, nums2_left_max)
            elif nums1_left_max > nums2_right_min:
                right = i - 1
            else:
                left = i + 1

        return 0


# Time Complexity: O(log(min(n1, n2))), where n1 and n2 are the lengths of the two input arrays.
# This is because we are performing a binary search on the smaller array.
# The binary search will take at most log(min(n1, n2)) iterations.
# Space Complexity: O(1)

if __name__ == "__main__":
    solution = MedianOfTwoSortedArrays()

    print("Expected Output: 2")
    print("Actual Output:", solution.findMedianSortedArrays([1, 3], [2]))

    print("Expected Output: 2.5")
    print("Actual Output:", solution.findMedianSortedArrays([1, 2], [3, 4]))
